### ISSUE TITLE
Eliot + Jupyter

### ISSUE THREAD

**Issue by timothyb0912** [itamarst/eliot#457]
Hi @itamarst, thanks for this great package!

I've recently got interested in using Eliot, especially in using it programmatically and through Jupyter notebooks.

I saw [here](https://pythonspeed.com/articles/logging-for-scientific-computing/) in the "The future: Eliot + Jupyer" section, that you were interested in others exposing Eliot logs through Jupyter. Are you still interested in this?

I have a proof of concept in this [gist](https://gist.github.com/timothyb0912/889042d3700411924a57bdc7e48ca8f4).
It uses `treelib` to draw tree-diagrams of the logs, based on the [eliottree](https://github.com/jonathanj/eliottree) package.

Is this the general direction that you'd like to see Eliot go?
I'm happy to add tests, open a PR to add this functionality, and work with you to get it done.
Just let me know, and thanks again.

**Comment by itamarst** [comment]
Hi,

Glad you're interested, and finding it useful.

You cou can by the way use `eliot.parse` for parsing; it also supports missing messages, and out-of-order messages, which can become useful when doing threaded or multi-process tracing. And I wonder if eliottree would work as a library; not sure author considered it as use-case, but you can ask.

I'd love to hear about your use case in more detail:

* My thought during that talk was that Eliot can complement Jupyter: Jupyter is great at exposing internal implementation details _so long as code is in a Jupyter cell_. As soon as you hit a library (or you're not using Jupyter), you no longer have that visibility. So my vague notion was "dump to disk, then explore in Jupyter".
* What you're doing is probably going beyond what I was doing, in that you're using it in Jupyter code where you could in theory just be doing `display()`. And... now that I see it being done, that does seem like it might be useful, e.g. once you get complex recursion or repetition, or when you want to delay reporting information.

So, yes, this does seem interesting. However, the idioms and requirements are not clear to me yet.

As a starting point, I suggest you just go and do it for whatever you real-world goals are (and I'd love to hear them), and when you hit roadblocks ask here, and we can figure it out. And then the output might be some features, or bugfixes, or documentation, or all of them.

Does that make sense?

**Comment by timothyb0912** [comment]
Yeah, definitely. I'll make use of Eliot for my current projects, and when I'm done, I'll have a more detailed set of feature requests, usage notes, and possibly bugfixes to share. These learnings can be used to figure out how best to expose Eliot's logs in Jupyter notebooks.

Here's my use case. I work as a data scientist, and at the completion of any proof of concept, I often have a collection of untested Jupyter notebooks. Alternatively, I inherit the legacy code of others, in a notebook or outside of it. At either point, I'll begin using test driven-development to re-write the code in a well-designed manner or simply to learn how it works for myself. However, I'll want to be able to ensure that my new code preserves the functionality of my old code.

One way to do this is to execute many examples with both the original and new code to ensure equality. If differences are discovered it indicates missing specifications / tests for my new code or bugs for the old code. Assuming the original code was correct, we'll want to add tests that guide us to remove the differences in output between the old and new code. Ideally, the test to be added would revolve around the comparison of the first internally computed value (as logged by Eliot) that exists but differs in the two code bases, given identical inputs.

Does this make sense? I want to use legacy code + examples + Eliot as a test oracle for the creation of test-driven-code that replaces the legacy code.

**Comment by itamarst** [comment]
Ooh, neat. For one of my consulting clients we were talking about a similar problem, seeing if new version of e.g. Pandas broke code, and I did a little prototyping (and very briefly considered Eliot) but we didn't end up going down that route. Interested to hear how it works for you.

So yeah, makes total sense.

I believe you should be able to do:

```python
from eliot.parse import Parser

tasks = list(Parser.parse_stream(your_list_of_messages))
```
